### PR TITLE
Dev 2607 multiselect

### PR DESCRIPTION
**Pull request by petrican**

### CODE REVIEW

File: src/components/widget/List/List.js [hunk 1]
```diff
@@ -349,7 +349,7 @@ ListWidget.propTypes = {
   mainProperty: PropTypes.object,
   isModal: PropTypes.bool,
   autoFocus: PropTypes.bool,
-  selected: PropTypes.object,
+  selected: PropTypes.oneOfType([PropTypes.object, PropTypes.array]),
```
Comment (siemiatj) [inline src/components/widget/List/List.js]: I wish we could force this to always be an array. But probably without types and with our current test coverage it'd be a nightmare :)

File: src/components/widget/MultiSelect.js [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+    };
+  }
+
+  componentDidMount() {
+    this.props.onFocus();
+  }
+
+  static getDerivedStateFromProps(nextProps) {
+    let selected =
+      nextProps.selectedItems && Array.isArray(nextProps.selectedItems)
```
Comment (siemiatj) [inline src/components/widget/MultiSelect.js]: can we not have nested ternaries plx ?

File: src/components/widget/RawWidget.js [hunk 3]
```diff
@@ -690,9 +697,45 @@ export class RawWidget extends Component {
             viewId={viewId}
             autoFocus={autoFocus}
             validStatus={widgetData[0].validStatus}
+            isMultiselect={isMultiselect}
           />
         );
 
+      case 'MultiListValue':
+        return (
+          <List
```
Comment (siemiatj) [inline src/components/widget/RawWidget.js]: if the only difference is the `isMultiselect` prop, maybe we can merge this into `listWidgetProps` and use for both single and multi select lists ?

File: src/components/widget/MultiSelect.js [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+      });
+      return {
+        checkedItems: updatedCheckedItems,
+      };
+    }
+
+    return null;
+  }
+
+  selectItem = (key, caption) => {
```
Comment (siemiatj) [inline src/components/widget/MultiSelect.js]: Can you add a comment on what's going on over here ? Why are we transforming `state.checkedItems` to string and then parsing it ?

File: src/components/widget/MultiSelect.js [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+    let selected = null;
+    let newCheckedItems = JSON.parse(JSON.stringify(this.state.checkedItems));
+    if (typeof this.state.checkedItems[key] === 'undefined') {
```
Comment (siemiatj) [inline src/components/widget/MultiSelect.js]: Maybe we could do something like :
```javascript
_.set(checkedItems, ['key1'], { key: 'key1', value: !_.get(checkedItems, ['key1', 'value'], true)});
```

File: src/components/widget/MultiSelect.js [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+    selected = Object.keys(newCheckedItems).map((item) => {
+      if (newCheckedItems[item].value === true) {
+        return {
+          key: newCheckedItems[item].key,
+          caption: newCheckedItems[item].caption,
+        };
+      }
+    });
+
+    this.props.onSelect(selected.filter((el) => typeof el !== 'undefined'));
```
Comment (siemiatj) [inline src/components/widget/MultiSelect.js]: Just wondering if we shouldn't call this after we've updated the state ? Or it won't make a difference here ?

File: src/components/widget/List/List.js (hunk 1 above)
Comment (petrican) [reply src/components/widget/List/List.js]: When is solo item selection (from dropdown) is an object.   When the selection comes from the multiselect si an array of objects

File: src/components/widget/RawWidget.js (hunk 3 above)
Comment (petrican) [reply src/components/widget/RawWidget.js]: This I believe is wise to be done after the BE will provide only the widget type 'MultiListValue'. For now let's leave it as it is pls.

File: src/components/widget/MultiSelect.js (hunk 6 above)
Comment (petrican) [reply src/components/widget/MultiSelect.js]: calling it here will guarantee the top filter is set accordingly. The sync to state is done in getDerivedStateFromProps

File: src/components/widget/MultiSelect.js (hunk 5 above)
Comment (petrican) [reply src/components/widget/MultiSelect.js]: No action should be done to directly alter the state. That is the reason I deep cloned the newCheckedItems in the first place.

Comment (petrican) [reply src/components/widget/MultiSelect.js]: Also didn't want to . add lodash and preferred the vanillaJS way :). Thanks for understanding .

File: src/components/widget/MultiSelect.js (hunk 2 above)
Comment (petrican) [reply src/components/widget/MultiSelect.js]: Yeah that I missed. Will fix asap. Thanks

Comment (petrican) [reply src/components/widget/MultiSelect.js]: Fixed.

**Pull request merged**